File: app/src/test/java/net/mbonnin/arcanetracker/TestParser.java
package net.mbonnin.arcanetracker;

import net.mbonnin.arcanetracker.parser.Entity;
import net.mbonnin.arcanetracker.parser.Game;
import net.mbonnin.arcanetracker.parser.GameLogic;
import net.mbonnin.arcanetracker.parser.PowerParser;
import net.mbonnin.hsmodel.CardId;
import net.mbonnin.hsmodel.CardJson;
import net.mbonnin.hsmodel.PlayerClass;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;

import timber.log.Timber;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class TestParser {

    static class SimpleListener implements GameLogic.Listener {
        public Game game;

        @Override
        public void gameStarted(Game game) {
            this.game = game;
        }

        @Override
        public void gameOver() {

        }

        @Override
        public void somethingChanged() {

        }
    }

    @BeforeClass
    public static void beforeClass() {
        Timber.plant(new TestTree());
        CardJson.INSTANCE.init("enUS", new ArrayList<>());
    }

    private void runParser(String resource, GameLogic.Listener listener) throws IOException {
        GameLogic.get().addListener(listener);
        PowerParser powerParser = new PowerParser(tag -> GameLogic.get().handleRootTag(tag), null);
        InputStream inputStream = getClass().getResourceAsStream(resource);
        BufferedReader br = new BufferedReader(new InputStreamReader(inputStream));
        String line;

        powerParser.onPreviousDataRead();
        while ((line = br.readLine()) != null) {
            powerParser.onLine(line);
        }
        GameLogic.get().removeListener(listener);
    }

    private Game runParser(String resource) throws IOException {
        SimpleListener listener = new SimpleListener();

        runParser(resource, listener);
        return listener.game;
    }

    @Test
    public void testCreatedBy() throws Exception {
        Game game = runParser("/created_by.log");

        Assert.assertEquals(game.findEntitySafe("73").extra.createdBy, CardId.SERVANT_OF_KALIMOS);
        Assert.assertEquals(game.findEntitySafe("78").extra.createdBy, CardId.SERVANT_OF_KALIMOS);
        Assert.assertEquals(game.findEntitySafe("75").extra.createdBy, CardId.FROZEN_CLONE);
        Assert.assertEquals(game.findEntitySafe("76").extra.createdBy, CardId.FROZEN_CLONE);
        Assert.assertEquals(game.findEntitySafe("80").extra.createdBy, CardId.MIRROR_ENTITY);
    }

    @Test
    public void testDoubleSecret() throws Exception {

        Game game = runParser("/double_secret.log");

        Assert.assertFalse(game.victory);

    }

    @Test
    public void testSecrets() throws Exception {

        runParser("/secrets.log", new SimpleListener() {
            @Override
            public void somethingChanged() {
                if ("19".equals(game.gameEntity.tags.get(Entity.KEY_TURN))) {
                    Entity secretEntity = game.findEntityUnsafe("84");
                    Assert.assertFalse(secretEntity.extra.competitiveSpiritTriggerConditionHappened);
                    Assert.assertTrue(secretEntity.extra.otherPlayerHeroPowered);
                    Assert.assertTrue(secretEntity.extra.otherPlayerPlayedMinion);
                    Assert.assertTrue(secretEntity.extra.selfHeroDamaged);
                    Assert.assertTrue(secretEntity.extra.selfHeroAttacked);
                }
            }
        });
    }

    @Test
    public void testMirrorEntity() throws Exception {
        runParser("/mirror_entity.log", new SimpleListener() {
            @Override
            public void somethingChanged() {
                Entity kabalCrystalRunner = game.findEntitySafe("84");
                if (Entity.ZONE_PLAY.equals(kabalCrystalRunner.tags.get(Entity.KEY_ZONE))) {
                    Entity iceBarrier = game.findEntitySafe("41");
                    Assert.assertTrue(iceBarrier.extra.otherPlayerPlayedMinion);
                }
            }
        });
    }

    @Test
    public void testFrozenClone() throws Exception {
        runParser("/mirror_entity.log", new SimpleListener() {
            @Override
            public void somethingChanged() {
                Entity kabalCrystalRunner = game.findEntitySafe("84");
                if (Entity.ZONE_PLAY.equals(kabalCrystalRunner.tags.get(Entity.KEY_ZONE))) {
                    Entity iceBarrier = game.findEntitySafe("41");
                    Assert.assertTrue(iceBarrier.extra.otherPlayerPlayedMinion);
                }
            }
        });
    }

    @Test
    public void testExploreUngoro() throws Exception {
        runParser("/exploreUngoro.log", new SimpleListener() {
            @Override
            public void somethingChanged() {
                Entity e = game.findEntityUnsafe("99");
                if (e != null) {
                    Assert.assertTrue(e.CardID.equals(CardId.CHOOSE_YOUR_PATH));
                }
            }
        });
    }

    @Test
    public void testInterrupted() throws Exception {
        class InterruptedListener extends SimpleListener{
            public int gameOverCount;

            @Override
            public void gameOver() {
                gameOverCount++;
            }
        }

        InterruptedListener listener = new InterruptedListener();
        runParser("/interrupted.log", listener);

        Assert.assertTrue(listener.gameOverCount == 1);
    }

    @Test
    public void testEndedInAttack() throws Exception {
        class InterruptedListener extends SimpleListener{
            public int gameOverCount;

            @Override
            public void gameOver() {
                gameOverCount++;
            }
        }

        InterruptedListener listener = new InterruptedListener();
        runParser("/endedInAttack.log", listener);

        Assert.assertTrue(listener.gameOverCount == 1);
    }

    @Test
    public void testSpectator() throws Exception {
        class InterruptedListener extends SimpleListener{
            public int gameOverCount;

            @Override
            public void gameOver() {
                gameOverCount++;
            }
        }

        InterruptedListener listener = new InterruptedListener();
        runParser("/continuation.log", listener);

        /*
         * the first game is a continuation game and should not be detected
         */
        Assert.assertTrue(listener.gameOverCount == 1);
    }

    @Test
    public void testEffigy() throws Exception {
        runParser("/effigy.log", new SimpleListener() {
            @Override
            public void somethingChanged() {
                if ("11".equals(game.gameEntity.tags.get(Entity.KEY_TURN))) {
                    Entity secretEntity = game.findEntityUnsafe("18");
                    Assert.assertFalse(secretEntity.extra.selfPlayerMinionDied);
                }
            }
        });
    }

    @Test
    public void testNemsy() throws Exception {
        SimpleListener listener = new SimpleListener() {
            @Override
            public void gameStarted(Game game) {
                super.gameStarted(game);
                Assert.assertTrue(game.player.playerClass().equals(PlayerClass.WARLOCK));

            }
        };

        runParser("/nemsy.log", listener);
    }

}

File: app/src/test/java/net/mbonnin/arcanetracker/TestParser.kt
package net.mbonnin.arcanetracker

import kotlinx.io.streams.asInput
import net.hearthsim.kotlin.hslog.PowerParser
import net.mbonnin.arcanetracker.hslog.Game
import net.mbonnin.arcanetracker.hslog.GameLogic
import net.mbonnin.hsmodel.CardJson
import okhttp3.OkHttpClient
import okhttp3.Request
import org.junit.Assert
import org.junit.BeforeClass
import org.junit.Test
import timber.log.Timber
import java.io.BufferedReader
import java.io.File
import java.io.InputStream
import java.io.InputStreamReader

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */
class TestParser {

    internal open class SimpleListener : GameLogic.Listener {
        var game: Game? = null

        override fun gameStarted(game: Game) {
            this.game = game
        }

        override fun gameOver() {

        }

        override fun somethingChanged() {

        }
    }

    private fun runParser(inputStream: InputStream, listener: GameLogic.Listener) {
        GameLogic.get().addListener(listener)
        val powerParser = PowerParser({ tag -> GameLogic.get().handleRootTag(tag) }, null, null)
        val br = BufferedReader(InputStreamReader(inputStream))
        var line: String?

        System.out.println("Running...")
        while (true) {
            line = br.readLine()
            if (line == null) {
                break
            }
            powerParser.process(line, processGameTags = true)
        }
        GameLogic.get().removeListener(listener)
    }

    private fun runParser(logFileName: String, listener: GameLogic.Listener) {
        val client = OkHttpClient()
        val request = Request.Builder().url("https://raw.githubusercontent.com/HearthSim/hsreplay-test-data/master/data/${logFileName}").get().build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw Exception()
        }

        val inputStream = response.body()!!.byteStream()
        runParser(inputStream, listener)
    }

    @Test
    fun testSpectator() {
        class InterruptedListener : SimpleListener() {
            var gameOverCount: Int = 0

            override fun gameOver() {
                gameOverCount++
            }
        }

        val listener = InterruptedListener()
        runParser("spectator.log", listener)

        Assert.assertTrue(listener.gameOverCount == 1)
    }

    @Test
    fun testZayle() {
        class InterruptedListener : SimpleListener() {
            override fun gameStarted(game: Game) {
                super.gameStarted(game)

                System.out.println("GameLogicListener.isPlayerZayle(game)=${GameLogic.isPlayerZayle(game)}")
            }
            override fun gameOver() {
            }
        }

        val listener = InterruptedListener()
        runParser("zayle.log", listener)
    }

    companion object {

        @BeforeClass
        fun beforeClass() {
            Timber.plant(TestTree())
            val input = File("/home/martin/git/Arcane-Tracker/app/src/main/res/raw/cards.json").inputStream().asInput()
            CardJson.init("enUS", null, input)
        }
    }

}